api/src/routers/meals.js

```diff
@@ -3,16 +3,61 @@ import knex from "../database_client.js";
 
 const router = express.Router();
 
-// router to get all the meals
+// router to get all the meals with query parameters
 router.get("/", async (req, res, next) => {
   try {
-    const meals = await knex.from("meal");
+    let query = knex("meal");
+    if (req.query.maxPrice) {
+      const maxPrice = parseInt(req.query.maxPrice);
+      query = query.where("price", "<=", maxPrice);
+    }
+
+    if (req.query.availableReservations) {
+      const available = req.query.availableReservations === "true";
+      query = query
+        .leftJoin("reservation", "meal.id", "reservation.meal_id")
+        .groupBy("meal.id")
+        .select("meal.id", "meal.title", "meal.price", "meal.max_reservations")
+        .having(
+          `meal.max_reservations - COUNT(reservation.id) ${available ? ">" : "="} 0`
+        );
+    }
+
+    if (req.query.title) {
+      const title = req.query.title;
+      query = query.where("title", "like", `%${title}%`);
+    }
+
+    if (req.query.dateAfter) {
+      const dateAfter = req.query.dateAfter;
+      query = query.where("when", ">", dateAfter);
+    }
+
+    if (req.query.dateBefore) {
+      const dateBefore = req.query.dateBefore;
+      query = query.where("when", "<", dateBefore);
+    }
+
+    if (req.query.limit) {
+      const limit = parseInt(req.query.limit);
+      query = query.limit(limit);
+    }
```

Comment:
Looks good, but you are missing some validation of the request here :) Think about it from the users perspective. If you were using an endpoint, wouldn't you appreciate if the server told you what part of the query was wrong?

For example limit must be an integer, but we don't tell the user that the value you gave us is invalid.

Use error code 400 for these errors